docs/4.4/api-reference.md

```diff
@@ -11,57 +11,328 @@ Teleport is currently working on documenting our API.
-# ...
+
+$ tctl create -f api-admin.yaml
+$ mkdir -p certs
+$ tctl auth sign --format=tls --user=api-admin --out=certs/api-admin
 ```
 
-The above sequence should result in three PEM encoded files being generated:
-`auth.crt`, `auth.key`, and `auth.cas` (certificate, private key, and CA certs respectively).
+This should result in three PEM encoded files being generated in the `/certs` directory: `api-admin.crt`, `api-admin.key`, and `api-admin.cas` (certificate, private key, and CA certs respectively).
 
-Note: by default, tctl auth sign produces certificates with a relatively short lifetime.
-For production deployments, the --ttl flag can be used to ensure a more practical
-certificate lifetime.
+Move the `/certs` folder into your `/api-examples` folder.
 
-# gRPC APIs
+!!! Note
+    By default, `tctl auth sign` produces certificates with a relatively short lifetime.
+    For production deployments, the `--ttl` flag can be used to ensure a more practical
+    certificate lifetime. See our [Kubernetes Section](kubernetes-ssh.md#using-teleport-kubernetes-with-automation) for more information on automating the signing process for short lived certificates.
+
+## Go Client
+
+Add `client.go` into `/api-examples`.
+
+**client.go**
+
+```go
+{!examples/go-client/client.go!}
+```
+
+## Go Packages
+
+Copy the Teleport module's go.mod below into `/api-examples` and then run `go mod tidy` to slim it down to only what's needed for these api examples.
+
+```
+{!go.mod!}
+```
+
+## Main file
+
+Add this main file to your `/api-examples` folder. Now you can simply plug in the examples below and then run `go run .` to see them in action.
+
+**main.go**
+
+```go
+package main
+
+import (
+	"fmt"
+	"log"
+)
+
+func main() {
+	log.Printf("Starting teleport client...")
+	client, err := connectClient()
+	if err != nil {
+		log.Fatalf("Failed to create client: %v", err)
+	}
+}
+```
+
+## Roles
+
+[Roles](enterprise/ssh-rbac.md#roles) are used to define what resources and actions a user is allowed/denied access to. It's best to make your role's permissions as stringent as possible.
+
+**Retrieve role**
+
+```go
+role, err := client.GetRole("auditor")
+if err != nil {
+  return err
+}
+```
+
+**Create a new role**
+
+```go
+// create a new auditor role which has very limited permissions
+role, err := services.NewRole("auditor", services.RoleSpecV3{
+  Options: services.RoleOptions{
+    MaxSessionTTL: services.Duration(time.Hour),
+  },
+  Allow: services.RoleConditions{
+    Logins: []string{"auditor"},
+    Rules: []services.Rule{
+      // - resources: ['session']
+      //   verbs: ['list', 'read']
+      services.NewRule(services.KindSession, services.RO()),
+    },
+  },
+  Deny: services.RoleConditions{
+    // node_labels: '*': '*'
+    NodeLabels: services.Labels{"*": []string{"*"}},
+  },
+})
+if err != nil {
+  return err
+}
+
+if err = client.UpsertRole(ctx, role); err != nil {
+  return err
+}
+```
```

Comment:
For this and all other examples, add a sentence of two that explain what this API does. Maybe even add some examples that the user can relate to, like how this would be done with `tctl` right now.

You have some of this in the function comment, but clean it up and explain it.

Here is very rough example of how I would update it with. @benarent what do you think?

**Create Role**

You can use the `UpsertRole` RPC to programmatically create a new role. You may want to create a new role if:

* You want to write a program that can always ensure certain roles exist on your system and do not want to orchestrate `tctl` to do this.
* You want to dynamically create short lived roles.
* You want to dynamically create roles with fields filled that Teleport currently does not support.

For example, suppose you wanted to create a role for an auditor that could view all sessions, but could not access any servers.

Using `tctl`, you would first create a role that allows reading and listing of the `session` session resource like below. In addition, the user is explicitly denied access to all nodes in the `deny` block.

```bash
$ cat << EOF > /tmp/auditor-role.yaml
kind: role
version: v3
metadata:
  name: auditor
spec:
  options:
    max_session_ttl: 8h
  allow:
    rules:
    - resources: [session]
      verbs: [list, read]
  deny: {}
    node_labels: '*': '*'
EOF
$ tctl create -f /tmp/auditor-role.yaml
```

To do something similar with the API:

```go
role, err := services.NewRole("auditor", services.RoleSpecV3{
  Options: services.RoleOptions{
    MaxSessionTTL: services.Duration(time.Hour),
  },
  Allow: services.RoleConditions{
    Logins: []string{"auditor"},
    Rules: []services.Rule{
      services.NewRule(services.KindSession, services.RO()),
    },
  },
  Deny: services.RoleConditions{
    NodeLabels: services.Labels{"*": []string{"*"}},
  },
})
if err != nil {
  return err
}
if err = client.UpsertRole(ctx, role); err != nil {
This conversation was marked as resolved by Joerger
  return err
}
```

Reply:
My thought process was that if a user is looking to use the API, they will already have a good idea what they need it for. Otherwise they would be using the more straightforward and developed CLI tools. I also don't have a good enough understanding of all of the use cases a user might have, and it would take a bit of time to dig into that, as well as multiple more reviews. How should we proceed given the deadline today?

Reply:
I think what we have is good and we can share it if needed. But let's clean this up some more next week before we merge.

Reply:
@russjones  the example you provided for Create Role, is spot on. Let's try to build out similar examples for the rest of the docs. I can help with this next week. ( We can also do this with a follow up PR )

Reply:
@Joerger Take a look at Stripe docs for balance. Even though it's self explanatory what balance is, it's still explained first.

https://stripe.com/docs/api/balance

Reply:
One of the other really nice things about their docs is they document the data types for input and output. Really makes using the API simple to use by reducing guesswork.